tests/API/BasicCrudTest.php

```diff
@@ -2,58 +2,148 @@
+    abstract protected function getUpdateBody();
+
+    /**
+     * Assertions for the created entity.
+     *
+     * @param array $created_entity - Created entity.
+     *
+     * @return mixed
+     */
+    abstract protected function afterCreate($created_entity);
+
+    /**
+     * Assertions for the updated entity.
+     *
+     * @param array $updated_entity - Updated entity.
+     *
+     * @return mixed
+     */
+    abstract protected function afterUpdate($updated_entity);
+
+    /**
+     * BasicCrudTest constructor.
+     * Sets up environment and domain value.
+     */
+    public function __construct()
+    {
+        parent::__construct();
+        $this->env = $this->getEnv();
+        $this->domain = $this->env['DOMAIN'];
+    }
 
-    protected function getAll($client, $entity) {
-        return $client->getAll();
+    /**
+     * Stub "get all entities" method.
+     * Can be overridden by child classes for specific test cases.
+     *
+     * @param mixed $crud_api - API client from the child class
+     * @param array $created_entity - Created entity.
+     *
+     * @return mixed
+     */
+    protected function getAll($crud_api, $created_entity)
+    {
+        return $crud_api->getAll();
     }
 
-    protected function getId($entity) {
+    /**
+     * Get the unique identifier for the entity.
+     *
+     * @param array $entity - Entity array.
+     *
+     * @return mixed
+     */
+    protected function getId($entity)
+    {
         return $entity['id'];
     }
 
-    public function testAll() {
-
-        $client = $this->getApiClient();
+    /**
+     * All basic CRUD test assertions.
+     */
+    public function testAll()
+    {
 
-        $options = $client->getApiClient()->get()->getGuzzleOptions();
+        // Get the API client from the child class.
+        $crud_api = $this->getApiClient();
 
+        // Get and check that our options have been set correctly.
+        $options = $crud_api->getApiClient()->get()->getGuzzleOptions();
         $this->assertArrayHasKey('base_uri', $options);
         $this->assertEquals("https://$this->domain/api/v2/", $options['base_uri']);
 
-        $created = $client->create($this->getCreateBody());
-
-        $all = $this->getAll($client, $created);
-
-        $found = false;
-        foreach ($all as $value) {
-            if ($this->getId($value) === $this->getId($created)) {
-                $found = true;
-                break;
+        // Test a generic "create entity" method for this API client.
+        $created_entity = $crud_api->create($this->getCreateBody());
+        $created_entity_id = $this->getId($created_entity);
+        $this->afterCreate($created_entity);
+
+        // Test a generic "get entity" method.
+        $got_entity = $crud_api->get($created_entity_id);
+        $this->afterCreate($got_entity);
+
+        // Test a generic "get all entities" method for this API client.
+        $all_entities = $this->getAll($crud_api, $created_entity);
```

Comment:
why is this call passing the crud api instead of calling getAll over that api in the first place? like you do on the rest of the methods..

Reply:
Will set `$crud_api` as a property of the class.